ludwig/data/prompt.py

```diff
@@ -0,0 +1,178 @@
+from ludwig.models.retrieval import get_retrieval_model, RetrievalModel
+from ludwig.utils.fs_utils import get_default_cache_location
+from ludwig.utils.types import Series
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_ZERO_SHOT_PROMPT_TEMPLATE = """SAMPLE INPUT: {sample_input}
+
+USER: Complete the following task: {task}
+
+ASSISTANT:
+"""
+
+
+DEFAULT_FEW_SHOT_PROMPT_TEMPLATE = """Below is relevant context:
+
+CONTEXT: {context}
+
+The context is comprised of labeled samples whose embeddings were similar to that of the sample input. The labels in
+these samples could aid you in your final prediction. Given this context and no prior knowledge, follow the instructions
+below.
+
+SAMPLE INPUT: {sample_input}
+
+USER: Complete the following task: {task}
+
+ASSISTANT:
+"""
+
+
+def index_column(
+    retrieval_config: Dict[str, Any],
+    col_name: str,
+    dataset_cols: Dict[str, Series],
+    backend: "Backend",
+    split_col: Optional[Series] = None,
+) -> Tuple[RetrievalModel, str]:
+    """Indexes a column for sample retrieval via embedding index lookup.
+
+    This function indexes a column and saves the index artifact to disk. If an index name is provided as part of the
+    `retrieval_config`, then the index in the ludwig cache with the corresponding name will be loaded instead of being
+    built from scratch.
+
+    To prevent data leakage, a split column must be provided. This ensures that the retrieval model only ever fetches
+    samples from the training set.
+
+    To ensure that the index is usable even if the original DataFrame is not available, the columns used to build the
+    index are stored as part of the index.
+
+    All operations in this function are performed on pandas objects, which means that you may run out of memory if your
+    dataset is large.
+
+    Args:
+        retrieval_config (Dict[str, Any]): The retrieval config from the config object.
+        col_name (str): The name of the column to index.
+        dataset_cols (Dict[str, Series]): A dictionary mapping column names to their corresponding Series. `col_name`
+            must be a key in this dictionary. These columns are stored as part of the index to ensure that the index
+            is usable even if the original DataFrame is not available.
+        df_engine (DataFrameEngine): The engine used to compute the columns into pandas objects.
+        split_col (Optional[Series]): A column that indicates whether a sample is part of the training set. A sample
+            is in the training set if the value in this column is 0.
+    Returns:
+        Tuple[RetrievalModel, str]: A tuple containing the retrieval model and the name of the index.
+    """
+    retrieval_model = get_retrieval_model(
+        retrieval_config["type"],
+        model_name=retrieval_config["model_name"],
+    )
+
+    index_name = retrieval_config["index_name"]
+    index_cache_directory = os.path.join(get_default_cache_location(), "index")
```

Comment:
Use `fs_utils` instead of native Python filesystem API, so we can support s3 and other remote filesystems for the cache.